# Домашнее задание №2

Тема "Создание признакового пространства"

In [1]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models import Word2Vec
import sklearn

In [4]:
# Загрузим сохраненные данные из предыдущей домашней работы

combine_df = pd.read_pickle("df.pkl")
#combine_df.drop(columns=['tweet','clean_tweet','tweet_token','tweet_token_filtered'], inplace=True)
combine_df.head(3)

id  ...                                   tweet_lemmatized
0   1  ...  [father, dysfunctional, selfish, drag, kid, dy...
1   2  ...  [thanks, lyft, credit, use, cause, offer, whee...
2   3  ...                                  [bihday, majesty]

[3 rows x 8 columns]

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*   Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*   Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*   Исключим стоп-слова с помощью stop_words='english'.
*   Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [5]:
# Переводим tweet_lemmatized и tweet_stemmed из списков строк в строки, чтобы vectorizerы отработали
combine_df['tweet_stemmed_str'] = combine_df['tweet_stemmed'].apply(lambda x: ' '.join(x))
combine_df['tweet_lemmatized_str'] = combine_df['tweet_lemmatized'].apply(lambda x: ' '.join(x))

In [6]:
count_vectorizer = CountVectorizer(max_df=0.9, max_features = 1000, stop_words='english')

In [7]:
# Создаем the Bag-of-Words модель
BOW_count_stemmed = count_vectorizer.fit_transform(combine_df['tweet_stemmed_str'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(BOW_count_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  act  ...  yoyou  yoyour  yoyoy  yr  yummi
0        0        0       0        0    0  ...      0       0      0   0      0
1        0        0       0        0    0  ...      0       0      0   0      0
2        0        0       0        0    0  ...      0       0      0   0      0
3        0        0       0        0    0  ...      2       1      0   0      0
4        0        0       0        0    0  ...      0       0      0   0      0
...    ...      ...     ...      ...  ...  ...    ...     ...    ...  ..    ...
49154    0        0       0        0    0  ...      0       0      0   0      0
49155    0        0       0        0    0  ...      0       0      0   0      0
49156    0        0       0        0    0  ...      0       0      0   0      0
49157    0        0       0        0    0  ...      0       0      0   0      0
49158    0        0       0        0    0  ...      0       0      0   0      0

[49159 rows x 1000 columns]

In [8]:
# Создаем the Bag-of-Words модель
BOW_count_lemmatized = count_vectorizer.fit_transform(combine_df['tweet_lemmatized_str'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(BOW_count_lemmatized.toarray(), columns = feature_names)

able  absolutely  account  act  ...  yoyour  yoyoyou  yr  yummy
0         0           0        0    0  ...       0        0   0      0
1         0           0        0    0  ...       0        0   0      0
2         0           0        0    0  ...       0        0   0      0
3         0           0        0    0  ...       1        0   0      0
4         0           0        0    0  ...       0        0   0      0
...     ...         ...      ...  ...  ...     ...      ...  ..    ...
49154     0           0        0    0  ...       0        0   0      0
49155     0           0        0    0  ...       0        0   0      0
49156     0           0        0    0  ...       0        0   0      0
49157     0           0        0    0  ...       0        0   0      0
49158     0           0        0    0  ...       0        0   0      0

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*   Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*   Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*   Исключим стоп-слова с помощью stop_words='english'.
*   Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [9]:
tfidf_vectorizer = TfidfVectorizer(max_df=0.9, max_features = 1000, stop_words='english')

In [10]:
# Создаем the Bag-of-Words модель
BOW_tfidf_stemmed = tfidf_vectorizer.fit_transform(combine_df['tweet_stemmed_str'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(BOW_tfidf_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  ...    yoyour  yoyoy   yr  yummi
0      0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
1      0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
2      0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
3      0.0      0.0     0.0      0.0  ...  0.392399    0.0  0.0    0.0
4      0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
...    ...      ...     ...      ...  ...       ...    ...  ...    ...
49154  0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
49155  0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
49156  0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
49157  0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0
49158  0.0      0.0     0.0      0.0  ...  0.000000    0.0  0.0    0.0

[49159 rows x 1000 columns]

In [11]:
# Создаем the Bag-of-Words модель
BOW_tfidf_lemmatized = tfidf_vectorizer.fit_transform(combine_df['tweet_lemmatized_str'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(BOW_tfidf_lemmatized.toarray(), columns = feature_names)

able  absolutely  account  act  ...   yoyour  yoyoyou   yr  yummy
0       0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
1       0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
2       0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
3       0.0         0.0      0.0  0.0  ...  0.39173      0.0  0.0    0.0
4       0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
...     ...         ...      ...  ...  ...      ...      ...  ...    ...
49154   0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
49155   0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
49156   0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
49157   0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0
49158   0.0         0.0      0.0  0.0  ...  0.00000      0.0  0.0    0.0

[49159 rows x 1000 columns]

3. Натренируем gensim.models.Word2Vec модель на наших данных.
*   Тренировать будем на токенизированных твитах combine_df['tweet_token']
*   Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
*   Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [35]:
model = Word2Vec(combine_df['tweet_token'],
                 size=200,
                 window=5,
                 min_count=1, # 1 - модель не будет игнорировать редкие слова
                 sg=1,
                 hs=0,
                 negative=10,
                 workers=32,
                 seed=34)

In [36]:
#model.train(combine_df['tweet_token'], total_examples=model.corpus_count, epochs=20)
model.train(combine_df['tweet_token'], total_examples=combine_df['tweet_token'].shape[0], epochs=20)

(9809186, 11728960)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [37]:
model.most_similar('dinner')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `most_similar` (Method will be removed in 4.0.0, use self.wv.most_similar() instead).
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('thegafford', 0.5591181516647339),
 ('eastvillage', 0.5556226968765259),
 ('cookout', 0.552168071269989),
 ('tacotuesday', 0.5521166324615479),
 ('spaghetti', 0.5510152578353882),
 ('roamingaround', 0.5507547855377197),
 ('nighout', 0.5451095700263977),
 ('bolognese', 0.5417038202285767),
 ('applebees', 0.5402548313140869),
 ('bihdaydinner', 0.5396629571914673)]

In [38]:
model.most_similar('trump')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `most_similar` (Method will be removed in 4.0.0, use self.wv.most_similar() instead).
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('donald', 0.6062121391296387),
 ('donaldtrump', 0.5793864727020264),
 ('suppoer', 0.5704876184463501),
 ('unfit', 0.5622063279151917),
 ('melo', 0.551409900188446),
 ('dumptrump', 0.5474816560745239),
 ('phony', 0.542633056640625),
 ('cuck', 0.5401263236999512),
 ('impeachment', 0.538715124130249),
 ('trumptrain', 0.5340621471405029)]

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".


In [39]:
model['food']

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


array([ 0.09090769, -0.21424697,  0.6012645 ,  1.0541756 ,  0.12455213,
       -0.17734484, -0.32768238, -0.5301158 , -0.52603674,  0.442806  ,
       -0.9621115 , -1.3125598 ,  0.17118308,  0.42067996,  0.17180459,
       -0.19799155, -0.23304644,  0.2183447 ,  0.26063988, -0.12263144,
       -0.03455845,  0.2643378 , -0.44827843,  0.53529555,  0.14139627,
       -0.5412488 , -0.85052174,  0.2074647 , -0.15974583, -0.86221147,
        0.30522737,  0.329499  , -0.19358146,  0.5916969 ,  0.24965976,
       -0.16807805, -0.15912436,  0.73713434, -1.1652315 ,  0.12701282,
        0.57951456, -0.3532627 ,  0.37865978, -0.42556238, -0.04629374,
       -0.59569645,  0.49608138,  0.15962817, -0.7125861 , -0.581939  ,
        0.02545026, -0.36278984, -0.37132773, -0.3047292 ,  0.17639723,
        0.5772706 ,  0.1329938 , -0.22870354, -0.31660298, -0.27792302,
        0.20041274,  0.12339849, -0.86899793, -0.761277  ,  0.5849657 ,
        0.8133514 ,  0.01461216,  0.70185477, -0.36998543, -0.44

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).


In [49]:
def get_tweet_vector(model, tokens):
  vec = 0
  if len(tokens) == 0:
    return 0
  for token in tokens:
    vec += model[token].reshape((1, model.vector_size))

  vec /= len(tokens)
  return vec

In [50]:
combine_df['tweet_vector'] = combine_df['tweet_token'].apply(lambda x: get_tweet_vector(model, x))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  


In [51]:
combine_df['tweet_vector']

0        [[0.051627908, 0.24656333, 0.10075648, 0.35455...
1        [[0.049836893, 0.19420157, 0.072499394, 0.1531...
2        [[-0.06390946, 0.5637483, -0.1860482, 0.041046...
3        [[0.07350253, 0.44853908, -0.104040414, 0.2799...
4        [[-0.0116906315, -0.019366957, -0.13190812, 0....
                               ...                        
49154    [[-0.03226529, 0.42374438, -0.2765781, 0.04882...
49155    [[0.2372659, 0.18005505, -0.037173223, 0.36937...
49156    [[0.12886684, 0.123452626, -0.08456679, 0.2686...
49157    [[-0.08258659, 0.3436687, -0.086908884, -0.138...
49158    [[0.052836273, 0.43060744, -0.013472631, 0.387...
Name: tweet_vector, Length: 49159, dtype: object

In [52]:
# Размерность полученных векторов
combine_df['tweet_vector'][0].shape

(1, 200)

In [54]:
# Общее количество твитов
combine_df['tweet_vector'].shape

(49159,)